design/design-doc-composition-functions-extra-resources.md

```diff
@@ -0,0 +1,1196 @@
+  map<string, string> match_labels = 4;
+}
+
+message RunFunctionRequest {
+  // Existing fields omitted for brevity
+
+  // Note that extra resources is a map to Resources, plural.
+  // The map key corresponds to the key in a RunFunctionResponse's
+  // extra_resources field. If a Function requests extra resources that
+  // don't exist Crossplane sets the map key to an empty Resources
+  // message to indicate that it attempted to satisfy the request.
+  map<string, Resources> extra_resources = 3;
+}
+
+// Resources just exists because you can't have map<string, repeated Resource>.
+message Resources {
+  repeated Resource resources = 1;
+}
+```
+
+And the logic according to which Crossplane will run the pipeline would be as
+follows:
+```golang
+// The following code should not be considered as a final implementation, but
+// rather as a pseudo-code to show the intended logic.
+
+// Compose resources using the Functions pipeline.
+func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructured, req CompositionRequest) (CompositionResult, error) {
+	// Existing code omitted for brevity...
+
+	// the set of watched resources, will be stored in `status.watchedResourceRefs`,
+	// will need to be converted in a sorted list before storing it
+	var resourcesToWatch map[string]v1beta1.ResourceRef
+
+	// Run any Composition Functions in the pipeline.
+	for _, fn := range req.Revision.Spec.Pipeline {
+		
+		// requestedResources is a map of allowed extra resources by name
+		// will be populated with the fetched resources and used to validate
+		// the watch requests
+		var requestedResources map[string]bool
+		
+		// used to store the resources fetched at the previous iteration
+        var extraResources map[string]v1beta1.Resources
+		
+		for i := int64(0); i < *fn.MaxIterations; i++ {
+			req := &v1beta1.RunFunctionRequest{Observed: o, Desired: d, Context: fctx, extraResources: extraResources}
+
+			// Run the Composition Function and get the response.
+			rsp, err := c.pipeline.RunFunction(ctx, fn.FunctionRef.Name, req)
+			if err != nil {
+				return CompositionResult{}, errors.Wrapf(err, /* ... */)
+			}
+
+			// Pass the desired state and context returned by this Function to the next iteration.
+			d = rsp.GetDesired()
+			fctx = rsp.GetContext()
+
+			rs := rsp.GetRequests()
+			// nothing to do if there are no requests to be satisfied
+			if rs == nil {
+				break
+			}
+
+			extraResourcesRequests := rs.GetExtraResources()
+
+			// if there are still extra resources getting requested on the last iteration, error out
+			if len(extraResourcesRequests) > 0 && i == *fn.MaxIterations-1 {
+				return CompositionResult{}, errors.Errorf( /* ... */ )
+			}
+
+			// check the references are in the observed state we just sent to the Function
+			// if not, error out
+			// otherwise add them to resourcesToWatch
+            ws := rs.GetWatchResources()
+			if err := checkAllowedWatchedResources(requestedResources, ws); err != nil {
```

Comment:
`maxIterations` set to 1 won't allow requesting any resource at all, as 1 means only a single request will be sent. Maybe we could make it default to 2 🤔 

`maxIterations` set to 2 will allow a Function to request some extra resource on the first iteration, getting them back at the second iteration and possibly requesting to watch them in the responses. This allows a Function to request a resource A and watch the same resource, but doesn't allow a Function to watch a resource B, e.g. a Secret, referenced by A, as the Function would need to request B before being able to watch it. I felt this was enough as if you care about just the reference to B, but not its content, you only actually want to watch A, while if you care about B's content, you will have to get it first and so you will be able to then watch it, and will need one more iteration.

Reply:
I wonder whether it makes sense for maxIterations to be configurable at all. Would it be enough to start with some sane default, just to avoid infinite iteration/retries?

Reply:
I agree that as a Composition Author picking a low value for `maxIterations` could be a hard task. So we could definitely pick a sane default and postpone providing such knob to a later iteration (😅)